File: 0044-wildcard-matching/wildcard-matching.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int i = 0, j = 0;
        int starIndex = -1;
        int matchIndex = 0;

        while (i < s.size()) {
            // Characters match, or '?' matches any one character
            if (j < p.size() && (p[j] == s[i] || p[j] == '?')) {
                i++;
                j++;
            }
            // Remember the position of '*'
            else if (j < p.size() && p[j] == '*') {
                starIndex = j;
                matchIndex = i;
                j++;
            }
            // Try expanding the previous '*'
            else if (starIndex != -1) {
                j = starIndex + 1;
                matchIndex++;
                i = matchIndex;
            }
            // No match possible
            else {
                return false;
            }
        }

        // Remaining pattern characters must all be '*'
        while (j < p.size() && p[j] == '*') {
            j++;
        }

        return j == p.size();
    }
};

File: 0044-wildcard-matching/wildcard-matching.py
class Solution:
    def isMatch(self, s, p):
        s_len, p_len = len(s), len(p)
        i = j = 0
        star = match = -1
        
        while i < s_len:
            if j < p_len and (p[j] == s[i] or p[j] == '?'):
                i += 1
                j += 1
            elif j < p_len and p[j] == '*':
                star = j
                match = i
                j += 1
            elif star != -1:
                j = star + 1
                match += 1
                i = match
            else:
                return False
        
        while j < p_len and p[j] == '*':
            j += 1
        
        return j == p_len